### PR TITLE
Update build to use gradle wrapper 6.8

### PR DESCRIPTION
**Pull request by breskeby**
- Updates our build to use the latest Gradle 6.8 release which is the last release before the major 7.0 release
- Resolve available gradle versions using built-in toolchain support
- Fixes deprecated usage of JavaInstallationRegistry
- We replace jdk handling in our build to rely on jvm detection provided by the gradle build tool itself. As we rely on environment variables pointing to jdks we wire this into the gradle jdk detection mechanism

### CODE REVIEW

**Comment by breskeby** [comment]
raised https://github.com/gradle/gradle/issues/15392 at gradle to handle the windows issues we see

**Comment by breskeby** [comment]
@elasticmachine test this please

File: buildSrc/src/test/java/org/elasticsearch/gradle/plugin/PluginBuildPluginTests.java [hunk 1]
```diff
@@ -1,73 +0,0 @@
...
-        extension.setClassname(getClass().getName());
-
-        ((ProjectInternal) project).evaluate();
-
-        assertNotNull(
-            "Task to generate notice not created: " + project.getTasks().stream().map(Task::getPath).collect(Collectors.joining(", ")),
-            project.getTasks().findByName("generateNotice")
-        );
-    }
-}
```
Comment (breskeby) [inline buildSrc/src/test/java/org/elasticsearch/gradle/plugin/PluginBuildPluginTests.java:73]: This test class is not working in a unit test context anymore. The only non ignored test checked for the existence of the registered existence. Thats not enough for me to have this ported to an integ tests at the moment so I removed the unit test.

**Comment by breskeby** [comment]
> Raised [gradle/gradle#15392](https://github.com/gradle/gradle/issues/15392) at gradle to handle the windows issues we see

That has been fixed in Gradle 6.8-rc-3

**Comment by breskeby** [comment]
@elasticmachine test this please

**Review by pugnascotia — APPROVED** [review]

**Review by mark-vieira — APPROVED** [review]
Few comments, otherwise LGTM.

File: buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java [hunk 2]
```diff
@@ -182,46 +187,33 @@ private JavaVersion determineJavaVersion(String description, File javaHome, Java
...
-        try {
-            return Files.isSameFile(javaHome.toPath(), Jvm.current().getJavaHome().toPath());
-        } catch (IOException e) {
-            throw new UncheckedIOException(e);
-        }
+    /**
+     * We resolve all available java versions using auto detected by gradles tool chain
+     * To make transition more reliable we only take env var provided installations into account for now
+     */
+    private List<JavaHome> getAvailableJavaVersions() {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:210]: good point. is that something you think is worth to invest in? I guess those issues are rare, but when they occur they suck to be debugged I can imagine. Maybe we want to look into this safety net in a follow up?

File: buildSrc/src/integTest/groovy/org/elasticsearch/gradle/fixtures/AbstractGradleFuncTest.groovy [hunk 3]
```diff
@@ -60,6 +63,10 @@ abstract class AbstractGradleFuncTest extends Specification {
         true
     }
 
+    def assertOutputContainsNot(String givenOutput, String expected) {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/integTest/groovy/org/elasticsearch/gradle/fixtures/AbstractGradleFuncTest.groovy:66]: thanks. `assertOutputMissing` it is

File: buildSrc/src/integTest/java/org/elasticsearch/gradle/BuildPluginIT.java [hunk 4]
```diff
@@ -105,7 +105,9 @@ private void runInsecureArtifactRepositoryTest(final String name, final String u
             .withProjectDir(tmpDir.getRoot())
             .withArguments("clean", "hello", "-s", "-i", "--warning-mode=all", "--scan")
             .withPluginClasspath()
+            .forwardOutput()
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/integTest/java/org/elasticsearch/gradle/BuildPluginIT.java:108]: I think we should keep it.

File: buildSrc/src/main/java/org/elasticsearch/gradle/precommit/JarHellPrecommitPlugin.java [hunk 5]
```diff
@@ -32,6 +32,9 @@
     public TaskProvider<? extends Task> createTask(Project project) {
         Configuration jarHellConfig = project.getConfigurations().create("jarHell");
         if (BuildParams.isInternal() && project.getPath().equals(":libs:elasticsearch-core") == false) {
+            // ideally we would configure this as a default dependency. But Default dependencies do not work correctly
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/precommit/JarHellPrecommitPlugin.java:35]: You would need to add a dependency like `jarHell 'org.elasticsearch:elasticsearch-core:7.10' at the moment. In general I think we need to put some more dedicated effort into making these external available plugins work nicely for contributors. By that I mean having proper test coverage of third party plugin author use cases etc. I remember we talked about that very early on when having our first talks about the elastic build already.  
We're moving in that direction with having internal plugins nowadays but there's more work required here

File: buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:210]: I think realistically what we _want_ to do is remove the reliance on these environment variables and use Gradle's toolchain capabilities to just download a compatible JDK when necessary. These variables are only used for _build_ time purposes, where we intend to run tests with a particular JDK implementation, that's what RUNTIME_JAVA_HOME is for.

So I think we can punt here and instead focus on ditching all usages of these things vs resolving JDKs dynamically.

[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:210]: I will look into this in a separate PR. I didn't want to make too many changes in this PR. So  `RUNTIME_JAVA_HOME` is basically used to configure the jdk on CI to run all tests against?

[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java:210]: The problem I see at the moment is that the compatibility choices (which vendor is available etc.) is limited in gradle at the moment and might not be enough. If you say that's less of an issue for `build time` then the toolchain capabilities in gradle might be enough at the moment to pick a proper one

**Pull request merged**